<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="description" content="OncoIMPACT : Cancer-omics Data Integration ">

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>OncoIMPACT</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/CSB5/oncoIMPACT">View on GitHub</a>

          <h1 id="project_title">OncoIMPACT</h1>
          <h2 id="project_tagline">Cancer-omics Data Integration </h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/CSB5/oncoIMPACT/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/CSB5/oncoIMPACT/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <hr>

<h2>
<a id="what-is-it" class="anchor" href="#what-is-it" aria-hidden="true"><span class="octicon octicon-link"></span></a>What is it?</h2>

<p>OncoIMPACT is a first-in-class algorithmic framework that nominates patient-specific driver genes by integratively modeling genomic mutations (point, structural and copy-number) and the resulting perturbations in transcriptional programs via defined molecular networks.</p>

<p>OncoIMPACT is configured to run in two modes: (1) a database mode that allows it to determine parameter settings from the data sets provided and (2) a discovery mode where information in the provided database is used to predict driver genes for each sample in an additional data set (which can be the same as the one used to create the database).</p>

<hr>

<h2>
<a id="installation" class="anchor" href="#installation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Installation</h2>

<p>Download and uncompress the latest version
    tar -zxvf oncoIMPACT.tgz
    chmod +x oncoIMPACT_v0.9.3/*.pl</p>

<hr>

<h2>
<a id="usage" class="anchor" href="#usage" aria-hidden="true"><span class="octicon octicon-link"></span></a>Usage</h2>

<p>Before running the pipeline, please ensure that you have the following files ready.</p>

<p><strong>Input file format</strong></p>

<ul>
<li><p>CNV_data: a tab-separated file with samples as columns and genes as rows. Each entry of this matrix must hold either a value of -1/0/1. A value of -1 would indicate a DELETION event while a value of 1 will indicate an AMPLIFICATION event, and a value of 0 if the is not affected by copy number gene.</p></li>
<li><p>SNP_data (point mutations and short indels): a tab-separated file with samples as columns and genes as rows. Each entry of this matrix must hold either the binary values of 0/1. A value of 1 would indicate the presence of a snp in the gene of that particular sample while a value of 0 would indicate otherwise.</p></li>
<li><p>EXP_data: a tab-separated file with samples as columns and genes as rows. Each entry of this matrix would represent the log2 fold-change of the gene expression of the tumor compared to a normal control.</p></li>
</ul>

<p><strong>Config file</strong>
You will need to create a config file for your specific project. The config file needs to contain the following parameters (you may refer to the sample config file provided with the scripts under &lt;<em>Scripts Directory/sampleConfig.cfg</em>&gt;).</p>

<ul>
<li><p>outDir: Full path to destination folder</p></li>
<li><p>scriptDir: Full path to folder where oncoIMPACT is installed</p></li>
<li><p>numThreads: Number of threads to use</p></li>
<li><p>cnv: Full path to cnv data matrix</p></li>
<li><p>exp: Full path to expression data matrix</p></li>
<li><p>dataType: Flag for expression data type. Vvalid options: ARRAY (default), RNA_SEQ</p></li>
<li><p>snp: Full path to snp data matrix</p></li>
<li><p>dataBase: Full path to the pre-computed database (implies discovery mode only)</p></li>
<li><p>databaseExport: Full path where the database will be exported (implies database + discovery mode)</p></li>
<li><p>testMode: Boolean flag to toggle test mode (valid options: 0 / 1)</p></li>
</ul>

<p>When you are ready to run the OncoIMPACT pipeline, simply enter the following command
    oncoIMPACT.pl </p>

<hr>

<h2>
<a id="sample-data" class="anchor" href="#sample-data" aria-hidden="true"><span class="octicon octicon-link"></span></a>Sample data</h2>

<p>We have provided 1 sample datasets for to test the OncoIMPACT pipeline;</p>

<ul>
<li><p>Glioblastoma.tgz: Glioblastoma (GBM) TCGA dataset</p></li>
<li><p>Glioblastoma_database.tgz: pre-computed database of the TCGA GBM dataset</p></li>
<li><p>single_patient.tgz: data for a single single GBM patient (to test the discovery mode)</p></li>
</ul>

<hr>

<h2>
<a id="change-log" class="anchor" href="#change-log" aria-hidden="true"><span class="octicon octicon-link"></span></a>Change Log</h2>

<p><strong>v0.9.3</strong>:</p>

<ul>
<li><p>NEW: Enable the processing of RNA-seq data</p></li>
<li><p>NEW: Enable the construction of databases</p></li>
<li><p>NEW: Enable the discovery mode using a pre-computed database</p></li>
</ul>

<p><strong>v0.9.2</strong>:</p>

<ul>
<li><p>NEW: Option in configuration file to run OncoIMPACT in test mode which performs the simulation with fewer iterations and fixed seed. In this mode, OncoIMPACT should complete in less than 2 hours using a single thread.</p></li>
<li><p>NEW: Sanity checks to ensure validity of parameters provided by user</p></li>
<li><p>FIX: Improved disk space utilization</p></li>
<li><p>FIX: Improved compatibility with Mac OS.</p></li>
</ul>

<p><strong>v0.9.1</strong>:</p>

<ul>
<li><p>oncoIMPACT will now avoid reproducing the input files if COMPLETE_SAMPLES folder exists</p></li>
<li><p>fix for bugs introduced in last version</p></li>
</ul>

<hr>

<h2>
<a id="licensing" class="anchor" href="#licensing" aria-hidden="true"><span class="octicon octicon-link"></span></a>Licensing</h2>

<p>The MIT License (MIT)
Copyright (c) 2014 Genome Institute of Singapore</p>

<hr>

<h2>
<a id="citing-oncoimpact" class="anchor" href="#citing-oncoimpact" aria-hidden="true"><span class="octicon octicon-link"></span></a>Citing OncoIMPACT</h2>

<p>Bertrand <em>et. al.</em>. <strong>Patient-specific driver gene prediction and risk assessment through integrated network analysis of cancer omics profiles.</strong> <a href="http://nar.oxfordjournals.org/content/43/7/e44.long"><em>Nucleic Acid Research</em> 2015, 43 (7): e44</a></p>

<hr>

<h2>
<a id="contacts" class="anchor" href="#contacts" aria-hidden="true"><span class="octicon octicon-link"></span></a>Contacts</h2>

<p>If you want community-driven support, please visit the forum at <a href="https://sourceforge.net/p/oncoimpact/discussion/">https://sourceforge.net/p/oncoimpact/discussion/</a>.</p>

<p>If you have a bug to report, you may raise a ticket at <a href="https://sourceforge.net/p/oncoimpact/tickets/">https://sourceforge.net/p/oncoimpact/tickets/</a>.</p>

<p>If you have other questions or feedback, you may direct them to Denis Bertrand (<a href="mailto:bertrandd@gis.a-star.edu.sg">bertrandd@gis.a-star.edu.sg</a>) or Burton Chia (<a href="mailto:chiakhb@gis.a-star.edu.sg">chiakhb@gis.a-star.edu.sg</a>).</p>
      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">OncoIMPACT maintained by <a href="https://github.com/CSB5">CSB5</a></p>
        <p>Published with <a href="https://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

    

  </body>
</html>
